src/include/simeng/config/Expectations.hh

```diff
@@ -0,0 +1,361 @@
+#pragma once
+
+#include <math.h>
+
+#include <algorithm>
+#include <cassert>
+#include <climits>
+#include <fstream>
+#include <iostream>
+#include <queue>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <variant>
+
+#include "simeng/config/yaml/ryml.hh"
+
+namespace simeng {
+namespace config {
+
+/** An enum containing all supported data types that can be expected of a config
+ * option. */
+enum ExpectedType { Bool, Double, Float, Integer, String, UInteger, Valueless };
+
+/** String to represent a wildcard node key. */
+const std::string wildcard = "*";
+
+/** A class to hold the expectations of a specific config option. Each class
+ * is considered to be one node of a tree-like structure which maps onto the
+ * hierarchical YAML structure of the passed/generated config file. Each node
+ * can contain any number of children, each of which is another instance of the
+ * `expectationNode` class for another config option. The expectation placed on
+ * each config option can be defined as a type, a set of values to which it must
+ * belong, and a set of bounds it must lie between. A default value is also
+ * expected for the sake of default construction and generation of default
+ * config files. The values of such expectations are held within a
+ * `std::variant` which can hold one of the expected data types equivalent to
+ * that held in the `ExpectedType` enum. */
+class expectationNode {
+  using dataTypeVariant =
+      std::variant<bool, double, float, int64_t, std::string, uint64_t>;
+
+ public:
+  expectationNode(){};
+
+  ~expectationNode(){};
+
+  /** A templated function to allow for the creation of an `expectationNode`
+   * instance. The instance created is one with a value and a key. A key,
+   * default value, type, and a flag denoting whether the node is optional are
+   * provided to the underlying constructor. */
+  template <typename T, bool optional = false>
+  expectationNode create(std::string key, T defaultValue) {
+    dataTypeVariant defValVariant = defaultValue;
+    ExpectedType type = static_cast<ExpectedType>(defValVariant.index());
+    // The instance created must have a value so ensure the passed type reflects
+    // requirement
+    if (type == ExpectedType::Valueless) {
+      std::cerr << "[SimEng:ModelConfig] Cannot create a node with a default "
+                   "value and an expected "
+                   "type of `Valueless`"
+                << std::endl;
+      exit(1);
+    }
+    expectationNode node = expectationNode(key, type, optional);
+    node.setDefaultValue(defValVariant);
+    return node;
+  }
+
+  /** A templated function to allow for the creation of an `expectationNode`
+   * instance. The instance created is one with only a key. A key and a flag
+   * denoting whether the node is optional are provided to the underlying
+   * constructor. */
+  template <bool optional = false>
+  expectationNode create(std::string key) {
```

Comment:
Good spot

Reply:
Given that both `create` functions enforce a constraint on the creation of `expectationNode` and don't change the internal state of an `expectationNode` instance;they don't need to be instance methods. It will be better if they are both static and not tied to an instance. We can then create an `expectationNode` via the invocation `expectationNode::create<...>(....)`

Also since the main difference between the create definitions is that one has a default value and other one doesn't. It is better if functions are only templated on default value `type`.

```
template <typename T>
static expectationNode create(T defaultValue, std::string key, bool optional = false); // call this f1

static expectationNode create(std::string, bool optional = false); // call this f2

This way we now have two separate invocations i.e

1) expectationNode::create("") and expectationNode::create("", true) // These invocations call f1 always
2) expectationNode::create<T>(defVal, "") and expectationNode::create<T>(defVal, "", true) // These invocations call f2 always
3) expectationNode::create<T>("") and expectationNode<T>("", true) // These won't compile cause default argument is not given.

```

Reply:
I'm not understanding the motivation for the changes.

Reply:
Everythings a lot more clear after this now, thank you for the clarification. Made changes akin to that discussed offline for the first point and remove the bool in the template.